core/src/main/java/hudson/security/SecurityRealm.java

```diff
@@ -290,15 +290,49 @@ public void doLogout(StaplerRequest req, StaplerResponse rsp) throws IOException
         Authentication auth = SecurityContextHolder.getContext().getAuthentication();
         SecurityContextHolder.clearContext();
 
-        // reset remember-me cookie
-        Cookie cookie = new Cookie(ACEGI_SECURITY_HASHED_REMEMBER_ME_COOKIE_KEY,"");
+        String contextPath = req.getContextPath().length() > 0 ? req.getContextPath() : "/";
+        resetRememberMeCookie(req, rsp, contextPath);
+        clearStaleSessionCookies(req, rsp, contextPath);
+
+        rsp.sendRedirect2(getPostLogOutUrl(req,auth));
+    }
+
+    private void resetRememberMeCookie(StaplerRequest req, StaplerResponse rsp, String contextPath) {
+        Cookie cookie = new Cookie(ACEGI_SECURITY_HASHED_REMEMBER_ME_COOKIE_KEY, "");
         cookie.setMaxAge(0);
         cookie.setSecure(req.isSecure());
         cookie.setHttpOnly(true);
-        cookie.setPath(req.getContextPath().length()>0 ? req.getContextPath() : "/");
+        cookie.setPath(contextPath);
         rsp.addCookie(cookie);
+    }
 
-        rsp.sendRedirect2(getPostLogOutUrl(req,auth));
+    private void clearStaleSessionCookies(StaplerRequest req, StaplerResponse rsp, String contextPath) {
+        /* While "executableWar.jetty.sessionIdCookieName" and
+         * "executableWar.jetty.disableCustomSessionIdCookieName"
+         * <https://github.com/jenkinsci/extras-executable-war/blob/6558df699d1366b18d045d2ffda3e970df377873/src/main/java/Main.java#L79-L97>
+         * can influence the current running behavior of the generated session cookie, we aren't interested
+         * in either of them at all.
+         *
+         * What matters to us are any stale cookies.
+         * Those cookies would have been created by this jenkins in a different incarnation, when it
+         * could, perhaps, have had different configuration flags, including for those configurables.
+         *
+         * Thus, we unconditionally zap all JSESSIONID. cookies.
+         * a new cookie will be generated by sendRedirect2(...)
+         *
+         * We don't care about JSESSIONID cookies outside our path because it's the browser's
+         * responsibility not to send them to us in the first place.
+         */
+        final String cookieName = "JSESSIONID.";
+        for (Cookie cookie : req.getCookies()) {
+            if (cookie.getName().startsWith(cookieName)) {
+                // one reason users log out is to clear their session(s)
```

Comment:
This may satisfy @oleg-nenashev 's concerns
```suggestion
                LOGGER.log(Level.FINE, "Removing cookie {0} during logout", cookie.getName());
                // one reason users log out is to clear their session(s)
```

also needs this import
```
import java.util.logging.Level;
```